cabot/plugin_test_utils.py

```diff
@@ -0,0 +1,230 @@
+            email='dolores@affirm.com',
+            is_active=True
+        )
+        self.fallback_officer = User.objects.create(
+            username='teddy@affirm.com',
+            password='fakepassword',
+            email='teddy@affirm.com',
+            is_active=True
+        )
+
+        self.schedule = Schedule.objects.create(name='Test Schedule', fallback_officer=self.fallback_officer)
+        Shift.objects.create(user=self.duty_officer,
+                             start=timezone.now() - timezone.timedelta(minutes=1),
+                             end=timezone.now() + timezone.timedelta(hours=1),
+                             schedule=self.schedule)
+        self.service.schedules.add(self.schedule)
+
+        self.assertEquals(get_duty_officers(self.schedule), [self.duty_officer])
+        self.assertEquals(get_fallback_officers(self.schedule), [self.fallback_officer])
+
+        self.service.update_status()  # set initial status
+
+        # make sure all users have profiles
+        for user in self.user, self.duty_officer, self.fallback_officer:
+            UserProfile.objects.create(user=user)
+
+        super(PluginTestCase, self).setUp()
+
+    def run_checks(self, checks, from_service_status=None):
+        # type: (List[Tuple[StatusCheck, bool, bool]], Union[None, str]) -> None
+        """
+        Simulates running the given checks with the given results, then updates the service (triggering alerts).
+        :param checks: list of (check, succeeded, acked) tuples
+        :param from_service_status: specify the service status to transition from (service.old_overall_status), optional
+        """
+        # clear any previous results
+        StatusCheckResult.objects.all().delete()
+
+        time_per_run = timezone.timedelta(seconds=1)
+        start = timezone.now() - time_per_run * (len(checks) + 1)
+        finish = start + time_per_run
+
+        for check, succeeded, acked in checks:
+            result = StatusCheckResult(check=check, time=start, time_complete=finish, succeeded=succeeded, acked=acked)
+            result.save()
+
+            check.last_run = finish
+            check.save()
+
+            start += time_per_run
+            finish += time_per_run
+
+        if from_service_status:
+            self.service.overall_status = from_service_status
+
+        self.service.update_status()
+
+    def transition_service(self, old_status, new_status):
+        """
+        Set old and current service state (ignoring check states), then trigger alert.
+        """
+        self.service.old_overall_status = old_status
+        self.service.overall_status = new_status
+        self.service.last_alert_sent = None
+
+        # PASSING -> PASSING never calls alert() normally (see Service.update_status())
+        self.assertFalse(old_status == new_status and old_status == Service.PASSING_STATUS,
+                         msg='PASSING -> PASSING should never occur naturally.')
```

Comment:
??? if the service status doesn't change between two runs of update service it can  be passing -> passing?

Reply:
It can, but alert() isn't called in that case:
```
# Service.update_status():
if not (self.overall_status == Service.PASSING_STATUS and self.old_overall_status == Service.PASSING_STATUS):
            self.alert()
```
TBH, it doesn't matter, I was just being weirdly pedantic. I'll remove the assert.